api/src/main/java/org/apache/iceberg/types/Types.java

```diff
@@ -414,42 +415,96 @@ public int hashCode() {
 
   public static class NestedField implements Serializable {
     public static NestedField optional(int id, String name, Type type) {
-      return new NestedField(true, id, name, type, null);
+      return new NestedField(true, id, name, type, null, null);
     }
 
     public static NestedField optional(int id, String name, Type type, String doc) {
-      return new NestedField(true, id, name, type, doc);
+      return new NestedField(true, id, name, type, null, doc);
+    }
+
+    public static NestedField optional(int id, String name, Type type, Object defaultValue, String doc) {
+      return new NestedField(true, id, name, type, defaultValue, doc);
+    }
+
+    public static NestedField optionalWithDefault(int id, String name, Type type, Object defaultValue) {
+      return new NestedField(true, id, name, type, defaultValue, null);
     }
 
     public static NestedField required(int id, String name, Type type) {
-      return new NestedField(false, id, name, type, null);
+      return new NestedField(false, id, name, type, null, null);
     }
 
     public static NestedField required(int id, String name, Type type, String doc) {
-      return new NestedField(false, id, name, type, doc);
+      return new NestedField(false, id, name, type, null, doc);
+    }
+
+    public static NestedField required(int id, String name, Type type, Object defaultValue, String doc) {
+      validateDefaultValueForRequiredField(defaultValue, name);
+      return new NestedField(false, id, name, type, defaultValue, doc);
+    }
+
+    public static NestedField requiredWithDefault(int id, String name, Type type, Object defaultValue) {
+      validateDefaultValueForRequiredField(defaultValue, name);
+      return new NestedField(false, id, name, type, defaultValue, null);
     }
 
     public static NestedField of(int id, boolean isOptional, String name, Type type) {
-      return new NestedField(isOptional, id, name, type, null);
+      return new NestedField(isOptional, id, name, type, null, null);
     }
 
     public static NestedField of(int id, boolean isOptional, String name, Type type, String doc) {
-      return new NestedField(isOptional, id, name, type, doc);
+      return new NestedField(isOptional, id, name, type, null, doc);
+    }
+
+    public static NestedField of(int id, boolean isOptional, String name, Type type, Object defaultValue, String doc) {
+      return new NestedField(isOptional, id, name, type, defaultValue, doc);
+    }
+
+    private static void validateDefaultValueForRequiredField(Object defaultValue, String fieldName) {
+      Preconditions.checkArgument(defaultValue != null,
+          "Cannot create NestedField with a null default for a required field " +  fieldName);
```

Comment:
here, it is a required field and we want to make sure the defaultValue  is not null, because (based on the table [here](https://github.com/linkedin/iceberg/pull/63#discussion_r612030559) this not allowed. In the constructor itself, however, null defaultValue is valid and it means no-defaultValue